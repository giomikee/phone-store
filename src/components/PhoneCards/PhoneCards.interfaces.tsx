import type { Phone } from '../../interfaces';

export interface Props {
    phones: Phone[]
}