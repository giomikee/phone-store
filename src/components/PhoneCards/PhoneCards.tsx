import { Card, CardActionArea, CardContent, CardMedia, Grid, Typography } from '@mui/material';
import type { Props } from './PhoneCards.interfaces';
import { HEADINGS, PAGES, PHONE_ID_PLACEHOLDER, PHONES_GRID } from '../../constants';
import PhoneChips from '../PhoneChips/PhoneChips';
import { useNavigate } from 'react-router-dom';

const PhoneCards = ({ phones }: Props) => {
    const navigate = useNavigate();
    const goToPhonePage = (id: string) => navigate(PAGES.phone.replace(PHONE_ID_PLACEHOLDER, id));

    return (
        <>
            {
                phones.map(({id, brand, name, basePrice, imageUrl}, index) => 
                    <Grid 
                        key={`${id}_${index}`}  // Workaround for duplicated phones in API - XMI-RN13P5G
                        size={PHONES_GRID}
                    >
                        <Card>
                            <CardActionArea 
                                data-testid={`phone-card-action-area-${id}_${index}`}
                                onClick={() => goToPhonePage(id)}
                            >
                                <CardMedia
                                    component="img"
                                    image={imageUrl}
                                    height="300"
                                    alt={`${brand} ${name}`}
                                />
                                <CardContent>
                                    <PhoneChips
                                        brand={brand} 
                                        basePrice={basePrice} 
                                        brandChipStyle={{ 
                                            mr: '4px',
                                            mb: {
                                                xs: '4px',
                                                md: 0,
                                            }
                                        }}
                                    />
                                    <Typography 
                                        component={HEADINGS.h3}
                                        variant={HEADINGS.h6}
                                    >
                                        {name}
                                    </Typography>
                                    
                                </CardContent>
                            </CardActionArea>
                        </Card>
                    </Grid>
                )
            }
        </>
    )
}

export default PhoneCards