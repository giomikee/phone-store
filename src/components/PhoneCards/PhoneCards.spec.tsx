import { describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/vitest';
import PhoneCards from './PhoneCards';
import { useNavigate } from 'react-router-dom';

vi.mock('react-router-dom', async () => ({
    ...(await vi.importActual('react-router-dom')),
    useNavigate: vi.fn().mockReturnValue(vi.fn()),
}));

describe('Given PhoneCards component', () => { 
    const defaultProps = {
        phones: [
            {
                id: '1',
                brand: 'Brand A',
                name: 'Phone 1',
                imageUrl: 'https://example.com/phone1.jpg',
                basePrice: 999,
            },
            {
                id: '2',
                brand: 'Brand B',
                name: 'Phone 2',
                imageUrl: 'https://example.com/phone2.jpg',
                basePrice: 899,
            },
        ],
    };
    describe('When component is rendered', () => { 
        it('Then it should match the snapshot', () => {
            const { container } = render(<PhoneCards {...defaultProps} />);
            expect(container).toMatchSnapshot();
        });
    });

    describe('When first phone card is clicked', () => { 
        it('Then it should navigate to phone details page', () => {
            const navigate = useNavigate();

            render(<PhoneCards {...defaultProps} />);
            fireEvent.click(screen.getByTestId('phone-card-action-area-1_0'));
            

            expect(navigate).toHaveBeenCalledWith('/phones/1');
        });
    })
});