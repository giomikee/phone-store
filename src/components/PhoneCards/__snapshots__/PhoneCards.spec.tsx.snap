// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given PhoneCards component > When component is rendered > Then it should match the snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-6 MuiGrid-grid-md-3 css-19vcg0t-MuiGrid-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <button
        class="MuiButtonBase-root MuiCardActionArea-root css-coyn9m-MuiButtonBase-root-MuiCardActionArea-root"
        data-testid="phone-card-action-area-1_0"
        tabindex="0"
        type="button"
      >
        <img
          alt="Brand A Phone 1"
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-1kuwa16-MuiCardMedia-root"
          height="300"
          src="https://example.com/phone1.jpg"
        />
        <div
          class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <div
              class="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorPrimary MuiChip-filledPrimary css-1rdkf6a-MuiChip-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconSmall MuiChip-iconColorPrimary css-1umw9bq-MuiSvgIcon-root"
                data-testid="SmartphoneIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 1.01 7 1c-1.1 0-2 .9-2 2v18c0 1.1.9 2 2 2h10c1.1 0 2-.9 2-2V3c0-1.1-.9-1.99-2-1.99M17 19H7V5h10z"
                />
              </svg>
              <span
                class="MuiChip-label MuiChip-labelSmall css-eccknh-MuiChip-label"
              >
                Brand A
              </span>
            </div>
            <div
              class="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorSecondary MuiChip-filledSecondary css-3p8epo-MuiChip-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconSmall MuiChip-iconColorSecondary css-1umw9bq-MuiSvgIcon-root"
                data-testid="EuroIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15 18.5c-2.51 0-4.68-1.42-5.76-3.5H15l1-2H8.58c-.05-.33-.08-.66-.08-1s.03-.67.08-1H15l1-2H9.24C10.32 6.92 12.5 5.5 15 5.5c1.61 0 3.09.59 4.23 1.57L21 5.3C19.41 3.87 17.3 3 15 3c-3.92 0-7.24 2.51-8.48 6H3l-1 2h4.06c-.04.33-.06.66-.06 1s.02.67.06 1H3l-1 2h4.52c1.24 3.49 4.56 6 8.48 6 2.31 0 4.41-.87 6-2.3l-1.78-1.77c-1.13.98-2.6 1.57-4.22 1.57"
                />
              </svg>
              <span
                class="MuiChip-label MuiChip-labelSmall css-eccknh-MuiChip-label"
              >
                desde 999.00 €
              </span>
            </div>
          </div>
          <h3
            class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
          >
            Phone 1
          </h3>
        </div>
        <span
          class="MuiCardActionArea-focusHighlight css-1h5un5t-MuiCardActionArea-focusHighlight"
        />
      </button>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-6 MuiGrid-grid-md-3 css-19vcg0t-MuiGrid-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <button
        class="MuiButtonBase-root MuiCardActionArea-root css-coyn9m-MuiButtonBase-root-MuiCardActionArea-root"
        data-testid="phone-card-action-area-2_1"
        tabindex="0"
        type="button"
      >
        <img
          alt="Brand B Phone 2"
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-1kuwa16-MuiCardMedia-root"
          height="300"
          src="https://example.com/phone2.jpg"
        />
        <div
          class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <div
              class="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorPrimary MuiChip-filledPrimary css-1rdkf6a-MuiChip-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconSmall MuiChip-iconColorPrimary css-1umw9bq-MuiSvgIcon-root"
                data-testid="SmartphoneIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 1.01 7 1c-1.1 0-2 .9-2 2v18c0 1.1.9 2 2 2h10c1.1 0 2-.9 2-2V3c0-1.1-.9-1.99-2-1.99M17 19H7V5h10z"
                />
              </svg>
              <span
                class="MuiChip-label MuiChip-labelSmall css-eccknh-MuiChip-label"
              >
                Brand B
              </span>
            </div>
            <div
              class="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorSecondary MuiChip-filledSecondary css-3p8epo-MuiChip-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconSmall MuiChip-iconColorSecondary css-1umw9bq-MuiSvgIcon-root"
                data-testid="EuroIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15 18.5c-2.51 0-4.68-1.42-5.76-3.5H15l1-2H8.58c-.05-.33-.08-.66-.08-1s.03-.67.08-1H15l1-2H9.24C10.32 6.92 12.5 5.5 15 5.5c1.61 0 3.09.59 4.23 1.57L21 5.3C19.41 3.87 17.3 3 15 3c-3.92 0-7.24 2.51-8.48 6H3l-1 2h4.06c-.04.33-.06.66-.06 1s.02.67.06 1H3l-1 2h4.52c1.24 3.49 4.56 6 8.48 6 2.31 0 4.41-.87 6-2.3l-1.78-1.77c-1.13.98-2.6 1.57-4.22 1.57"
                />
              </svg>
              <span
                class="MuiChip-label MuiChip-labelSmall css-eccknh-MuiChip-label"
              >
                desde 899.00 €
              </span>
            </div>
          </div>
          <h3
            class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
          >
            Phone 2
          </h3>
        </div>
        <span
          class="MuiCardActionArea-focusHighlight css-1h5un5t-MuiCardActionArea-focusHighlight"
        />
      </button>
    </div>
  </div>
</div>
`;
